lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+          }
+        }
+      }
+    }
+  }
+}
+
+/// \brief Check that DO loops properly nest
+template<typename A>
+inline void CheckDoNesting(const A &loopBodies, ErrorHandler &eh) {
+  for (auto i1{loopBodies.cbegin()}; i1 != loopBodies.cend(); ++i1) {
+    const auto &v1{*i1};
+    for (auto i2{i1 + 1}; i2 != loopBodies.cend(); ++i2) {
+      const auto &v2{*i2};
+      assert(v1.first.begin() != v2.first.begin());
+      if ((v2.first.begin() < v1.second.end()) &&
+          (v1.second.begin() < v2.second.begin())) {
+        // this is an error: DOs do not properly nest
+        Report(eh, v2.second,
+            parser::MessageFormattedText{"'%s' doesn't properly nest"_err_en_US,
+                v1.first.ToString().c_str()});
+      }
+    }
+  }
+}
+
+/// \brief Advance \p Pos past any label and whitespace
+/// Want the statement without its label for error messages, range checking
+template<typename A> inline A SkipLabel(const A &position) {
+  const long maxPosition{position.end() - position.begin()};
+  if (maxPosition && (position[0] >= '0') && (position[0] <= '9')) {
+    long i{1l};
+    for (; (i < maxPosition) && std::isdigit(position[i]); ++i)
+      ;
+    for (; (i < maxPosition) && std::isspace(position[i]); ++i)
+      ;
+    return parser::CharBlock{position.begin() + i, position.end()};
+  }
+  return position;
+}
+
+/// \brief Check constraints on <i>label-do-stmt</i>
+template<typename A, typename B, typename C>
+inline void CheckLabelDoConstraints(const A &dos, const A &branches,
+    const B &labels, const C &scopes, ErrorHandler &eh) {
+  IndexList loopBodies;
+  for (const auto stmt : dos) {
+    const auto &label{std::get<parser::Label>(stmt)};
+    const auto &scope{std::get<ScopeProxy>(stmt)};
+    const auto &position{std::get<parser::CharBlock>(stmt)};
+    auto doTarget{GetLabel(labels, label)};
+    if (HasNoScope(std::get<ScopeProxy>(doTarget))) {
+      // C1133: this is an error: label not found
+      Report(eh, position,
+          parser::MessageFormattedText{
+              "label '%lu' cannot be found"_err_en_US, label});
+    } else if (std::get<parser::CharBlock>(doTarget).begin() <
+        position.begin()) {
+      // R1119: this is an error: label does not follow DO
+      Report(eh, position,
+          parser::MessageFormattedText{
+              "label '%lu' doesn't lexically follow DO stmt"_err_en_US, label});
+    } else if (!InInclusiveScope(
+                   scopes, scope, std::get<ScopeProxy>(doTarget))) {
+      // C1133: this is an error: label is not in scope
+      if (isStrictF18) {
+        Report(eh, position,
+            parser::MessageFormattedText{
+                "label '%lu' is not in scope"_err_en_US, label});
+      } else {
+        Report(eh, position,
+            parser::MessageFormattedText{
+                "label '%lu' is not in scope"_en_US, label});
+      }
+    } else if ((std::get<TargetStmtType>(doTarget) &
```

Comment:
Use `.test(TargetStatementEnum::Do)` when checking for a nonempty intersection with a singleton.